zipkin-server/src/main/java/zipkin2/server/internal/throttle/ThrottledCall.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2015-2019 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.server.internal.throttle;
+
+import com.netflix.concurrency.limits.Limiter;
+import com.netflix.concurrency.limits.Limiter.Listener;
+import java.io.IOException;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.function.Supplier;
+import zipkin2.Call;
+import zipkin2.Callback;
+import zipkin2.storage.InMemoryStorage;
+
+/**
+ * {@link Call} implementation that is backed by an {@link ExecutorService}. The ExecutorService serves two
+ * purposes:
+ * <ol>
+ * <li>Limits the number of requests that can run in parallel.</li>
+ * <li>Depending on configuration, can queue up requests to make sure we don't aggressively drop requests that would
+ *     otherwise succeed if given a moment. Bounded queues are safest for this as unbounded ones can lead to heap
+ *     exhaustion and {@link OutOfMemoryError OOM errors}.</li>
+ * </ol>
+ *
+ * @see ThrottledStorageComponent
+ */
+class ThrottledCall<V> extends Call<V> {
+  private final ExecutorService executor;
+  private final Limiter<Void> limiter;
+  private final Listener limitListener;
+  /**
+   * Delegate call needs to be supplied later to avoid having it take action when it is created (like
+   * {@link InMemoryStorage} and thus avoid being throttled.
+   */
+  private final Supplier<Call<V>> delegate;
+  private Call<V> call;
+  private boolean canceled;
+
+  public ThrottledCall(ExecutorService executor, Limiter<Void> limiter, Supplier<Call<V>> delegate) {
+    this.executor = executor;
+    this.limiter = limiter;
+    this.limitListener = limiter.acquire(null).orElseThrow(RejectedExecutionException::new);
+    this.delegate = delegate;
+  }
+
+  private ThrottledCall(ThrottledCall other) {
+    this(other.executor, other.limiter, other.call == null ? other.delegate : () -> other.call.clone());
+  }
+
+  @Override
+  public V execute() throws IOException {
+    try {
+      call = delegate.get();
+
+      // Make sure we throttle
+      Future<V> future = executor.submit(() -> {
+        try (AutoCloseable nameReverter = updateThreadName(call.toString())) {
```

Comment:
I went ahead and added naming to the Threads here (and in enqueue()) based on discussion from the workshop.  Let me know if this doesn't work for some reason.